game/view/gameplay/actionhud/init.lua

```diff
@@ -109,15 +105,29 @@ function ActionHUD:init(route)
   ADJACENCY.unset(self.adjacency)
 
   -- Control hints
-  self.cancel_hint = CancelHint(13, 8)
+  self.hand_hint = ControlHint(240+13, 28, "button-key-draw_hand", {
+    DEFAULT = "Open hand", FOCUS = "Close hand"
+  })
+  self.hand_hint:register("HUD_MIDDLE")
+  self.open_packs_hint = ControlHint(240+138, 8, "button-key-open_packs", {
+    DEFAULT = "See packs", FOCUS = "Discard"
+  })
+  self.open_packs_hint:register("HUD_MIDDLE")
+  self.cancel_hint = ControlHint(240+138, 48, "button-key-cancel", {
+    DEFAULT = "Wait", FOCUS = "Cancel"
+  })
   self.cancel_hint:register("HUD_MIDDLE")
-  self.confirm_hint = ConfirmHint(138, 8)
+  self.confirm_hint = ControlHint(240+263, 28, "button-key-confirm", {
+    DEFAULT = "Interact", FOCUS = "Play card"
+  })
   self.confirm_hint:register("HUD_MIDDLE")
-  self.open_packs_hint = OpenPacksHint(304, 8)
-  self.open_packs_hint:register("HUD_MIDDLE")
-  self.show_stats_hint = ShowStatsHint(527, 8)
+  self.show_stats_hint = ControlHint(13, 8, "button-key-show_stats", {
+    DEFAULT = "(hold) Show stats", FOCUS = "(hold) Show stats"
+  })
   self.show_stats_hint:register("HUD_MIDDLE")
-  self.toggle_hints_hint = ToggleHintsHint(734, 8)
+  self.toggle_hints_hint = ControlHint(13, 48, "button-key-toggle_hints", {
+    DEFAULT = "Toggle hints", FOCUS = "Toggle hints"
+  })
```

Comment:
these param values seem very arbitrary to me, especially because the last one is a godamn table.
wouldn't it be better to just pass a single table for all the params? this way all params would be _named_, and we would know all the required fields of the constructor

Reply:
That last argument is a table that maps a UI state name to what hint will be shown for that button icon. While I agree that it is by far not the best API (and that it is coupled to its internal implementation), I don't see how grouping all parameters in a single table would help in knowing which are required and which are not. They *would* be named, though, but that would also be inconsistent other constructor APIs.

Possible improvements we could do here:
+ accept a `vec2` parameter for the object position instead of a `(x, y)` "tuple" - this should make it clearer
+ maybe store icon file names in a table so the third parameters becomes something like `TOGGLE_HINT_ICON_ID`
+ alternatively, load it from the database, but that is a much larger refactoration
+ maybe use a sequence for the hints since the UI states are predetermined? that would require some constants for the indices, probably

What do you think?

Reply:
- i like the idea of storing the icon file name somewhere else. ideally it should be in the database, yeah, but just making it so it's not in a magical literal would improve this a lot.
- the vector instead of two numbers seems better too, but it does seem a bit overkill and would cause reference issues in the constructor itself
- grouping params in a single table would help because just by reading an example of usage of the API, i can easily understand what is going on, and although that in itself won't _completely_ tell me what params are required or not, they will most certainly tell me in a general mannter some or all of the params a function accepts